src/platform/audio/ffmpegaudiofilecache.cpp

```diff
@@ -0,0 +1,537 @@
+        }
+
+        return nullptr;
+    }
+
+    FFmpegOpenAudioFile open_audio;
+    open_audio.wave_data = static_cast<uint8_t *>(av_malloc(sizeof(WavHeader)));
+    open_audio.data_size = sizeof(WavHeader);
+    open_audio.audio_event_info = audio_event->Get_Event_Info();
+
+    if (!Open_FFmpeg_Contexts(&open_audio, file)) {
+        captainslog_warn("Failed to load audio file '%s', could not cache.", filename.Str());
+        return nullptr;
+    }
+
+    if (audio_event->Is_Positional_Audio() && open_audio.codec_ctx->channels > 1) {
+        captainslog_error("Audio marked as positional audio cannot have more than one channel.");
+        return nullptr;
+    }
+
+    if (!Decode_FFmpeg(&open_audio)) {
+        captainslog_warn("Failed to decode audio file '%s', could not cache.", filename.Str());
+        Close_FFmpeg_Contexts(&open_audio);
+        return nullptr;
+    }
+
+    Fill_Wave_Data(&open_audio);
+    Close_FFmpeg_Contexts(&open_audio);
+
+    open_audio.ref_count = 1;
+    m_currentSize += open_audio.data_size;
+
+    // m_maxSize prevents using overly large amounts of memory, so if we are over it, unload some other samples.
+    if (m_currentSize > m_maxSize && !Free_Space_For_Sample(open_audio)) {
+        captainslog_warn("Cannot play audio file since cache is full: %s", filename.Str());
+        m_currentSize -= open_audio.data_size;
+        Release_Open_Audio(&open_audio);
+
+        return nullptr;
+    }
+
+    m_cacheMap[filename] = open_audio;
+
+    return open_audio.wave_data;
+}
+
+/**
+ * Closes a file, reducing the references to it. Does not actually free the cache.
+ */
+void FFmpegAudioFileCache::Close_File(void *file)
+{
+    if (file == nullptr) {
+        return;
+    }
+
+    ScopedMutexClass lock(&m_mutex);
+
+    for (auto it = m_cacheMap.begin(); it != m_cacheMap.end(); ++it) {
+        if (it->second.wave_data == file) {
+            --it->second.ref_count;
+
+            break;
+        }
+    }
+}
+
+/**
+ * Sets the maximum amount of memory in bytes that the cache should use.
+ */
+void FFmpegAudioFileCache::Set_Max_Size(unsigned size)
+{
+    ScopedMutexClass lock(&m_mutex);
+    m_maxSize = size;
+}
+
+/**
+ * Attempts to free space by releasing files with no references
+ */
+unsigned FFmpegAudioFileCache::Free_Space()
+{
+    std::list<Utf8String> to_free;
+    unsigned freed = 0;
+
+    // First check for samples that don't have any references.
+    for (const auto &cached : m_cacheMap) {
+        if (cached.second.ref_count == 0) {
+            to_free.push_back(cached.first);
+            freed += cached.second.data_size;
+        }
+    }
+
+    for (const auto &file : to_free) {
+        auto to_remove = m_cacheMap.find(file);
+
+        if (to_remove != m_cacheMap.end()) {
+            Release_Open_Audio(&to_remove->second);
+            m_currentSize -= to_remove->second.data_size;
+            m_cacheMap.erase(to_remove);
+        }
+    }
+
```

Comment:
I question if this is the most efficient way of doing this. You are creating a list of items to remove doing allocations to create that list then remove them. It would be easier if you just did:
```cpp
// Free any unreferenced audio
for (const auto &cached : m_cacheMap) {
        if (cached.second.ref_count == 0) {
            freed += cached.second.data_size;
            Release_Open_Audio(&to_remove->second);
            m_currentSize -= to_remove->second.data_size;
        }
}
// Erase the unreferenced entries
for(auto it =m_cacheMap.begin(); it != m_cacheMap.end(); ) {
    if(it->second.ref_count == 0){
        it = m_cacheMap.erase(it);
    }
    else{
        ++it;
    }
}
```

Reply:
I think it is ok if FFmpegAudioCache is similar to MilesAudioCache for now. No urgent need to deviate just yet. Maybe both can be refactored at the same time.